pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/retention/RetentionManager.java

```diff
@@ -177,4 +193,63 @@ private boolean shouldDeleteInProgressLLCSegment(String segmentName, IdealState
       return states.size() == 1 && states.contains(CommonConstants.Helix.StateModel.SegmentStateModel.OFFLINE);
     }
   }
+
+  private void manageSegmentLineageCleanupForTable(String tableNameWithType) {
+    try {
+      DEFAULT_RETRY_POLICY.attempt(() -> {
+        // Fetch segment lineage
+        ZNRecord segmentLineageZNRecord = SegmentLineageAccessHelper
+            .getSegmentLineageZNRecord(_pinotHelixResourceManager.getPropertyStore(), tableNameWithType);
+        if (segmentLineageZNRecord == null) {
+          return true;
+        }
+        SegmentLineage segmentLineage = SegmentLineage.fromZNRecord(segmentLineageZNRecord);
+        int expectedVersion = segmentLineageZNRecord.getVersion();
+
+        // 1. The original segments can be deleted once the merged segments are successfully uploaded
+        // 2. The zombie lineage entry & merged segments should be deleted if the segment replacement failed in
+        //    the middle
+        Set<String> segmentsForTable = new HashSet<>(_pinotHelixResourceManager.getSegmentsFor(tableNameWithType));
+        List<String> segmentsToDelete = new ArrayList<>();
+        for (String lineageEntryId : segmentLineage.getLineageEntryIds()) {
+          LineageEntry lineageEntry = segmentLineage.getLineageEntry(lineageEntryId);
+          if (lineageEntry.getState() == LineageEntryState.COMPLETED) {
+            // If the lineage state is 'COMPLETED', it is safe to delete all segments from 'segmentsFrom'
+            segmentsToDelete.addAll(lineageEntry.getSegmentsFrom());
+
+            // The lineage entry with 'COMPLETED' state can only be safely removed when both segmentFrom & segmentTo
+            // are all removed from the table.
+            if (Collections.disjoint(segmentsForTable, lineageEntry.getSegmentsFrom()) && Collections
+                .disjoint(segmentsForTable, lineageEntry.getSegmentsTo())) {
+              segmentLineage.deleteLineageEntry(lineageEntryId);
+            }
```

Comment:
We can simplify the logic as following, same for `IN_PROGRESS`
```suggestion
            Set<String> sourceSegments = new HashSet<>(lineageEntry.getSegmentsFrom());
            sourceSegments.retainAll(segmentsForTable);
            if (sourceSegments.isEmpty()) {
              segmentLineage.deleteLineageEntry(lineageEntryId);
            } else {
              segmentsToDelete.addAll(sourceSegments);
            }
```

Reply:
Changed. Please double check the logic for `IN_PROGRESS` as well.